src/linear_algebra/affine_space/affine_subspace.lean

```diff
@@ -393,6 +393,74 @@ rfl
   ↑(a +ᵥ b) = (a:V) +ᵥ (b:P) :=
 rfl
 
+instance : has_vadd V (affine_subspace k P) :=
+{ vadd := λ x S,
+{ carrier := x +ᵥ S,
+  smul_vsub_vadd_mem := λ v _ _ _ ha hb hc,
+  begin
+    rcases ha with ⟨a, haS, rfl⟩,
+    rcases hb with ⟨b, hbS, rfl⟩,
+    rcases hc with ⟨c, hcS, rfl⟩,
+    simp only [affine_subspace.mem_coe, vadd_vsub_vadd_cancel_left] at *,
+    refine ⟨v • (a -ᵥ b) +ᵥ c, S.smul_vsub_vadd_mem v haS hbS hcS, _⟩,
+    rw [vadd_vadd, vadd_vadd, add_comm],
+  end } }
+
+instance : add_action V (affine_subspace k P) :=
+{ vadd := (+ᵥ),
+  zero_vadd := λ p, by { unfold has_vadd.vadd, simp only [zero_vadd, set.image_id'], ext, refl, },
+  add_vadd := λ x y p,
+  begin
+    unfold has_vadd.vadd,
+    simp only [set.image_vadd],
+    ext z,
+    rw [set.mem_vadd_set, set.mem_vadd_set],
+    refine ⟨λ h, _, λ h, _⟩,
+    { rcases h with ⟨a, ha, rfl⟩,
+      refine ⟨y +ᵥ a, ⟨a, ha, rfl⟩, vadd_vadd x y a⟩, },
+    rcases h with ⟨a, ha, rfl⟩,
+    refine ⟨-y +ᵥ a, _, _⟩,
+    { rcases ha with ⟨b, hb, rfl⟩,
+      rwa [vadd_vadd, add_left_neg, zero_vadd], },
+    rw [vadd_vadd, add_neg_cancel_right],
+  end }
+
+@[simp] lemma coe_const_vadd (v : V) (s : affine_subspace k P) :
+  ((v +ᵥ s : affine_subspace k P) : set P) = v +ᵥ s := rfl
+
+lemma mem_const_vadd_iff (v : V) {s : affine_subspace k P} {p : P} :
+  v +ᵥ p ∈ v +ᵥ s ↔ p ∈ s :=
+vadd_mem_vadd_set_iff
+
+lemma const_vadd_empty {v : V} {s : affine_subspace k P} (hs : (s : set P) = ∅) :
+  v +ᵥ s = s :=
+by { ext, simp [hs] }
+
+lemma mem_const_vadd_direction (v x : V) {s : affine_subspace k P} :
+  x ∈ s.direction → x ∈ (v +ᵥ s).direction :=
+begin
+  by_cases hem : set.nonempty (s : set P),
+  { rw mem_direction_iff_eq_vsub,
+    rintro ⟨p, hp, q, hq, rfl⟩,
+    rw ← vadd_vsub_vadd_cancel_left v _ _,
+    rw ← mem_const_vadd_iff v at hp,
+    rw ← mem_const_vadd_iff v at hq,
+    exact vsub_mem_direction hp hq,
+    exact hem, },
+  rw not_nonempty_iff_eq_empty at hem,
+  rw const_vadd_empty hem,
+  exact id,
+end
+
+lemma const_vadd_direction (v : V) (s : affine_subspace k P) :
+  (v +ᵥ s).direction = s.direction :=
+begin
+  ext, refine ⟨λ h, _, λ h, mem_const_vadd_direction v x h⟩,
+  have hx := mem_const_vadd_direction (-v) x h,
+  rwa neg_vadd_vadd at hx,
+end
+
+
```

Comment:
I think you should move this much further down the file - in particular, I think you will find it convenient to define this in terms of `map`

Reply:
In fact; lets put this in a new `pointwise.lean` file to match what we do elsewhere.

Reply:
Ok! I've done that now